### PR TITLE
align and cleanup Kokkos::complex

### PR DESCRIPTION
**Pull request by dsunder**
1.  added alignas(2*alignof(RealType)),
  1.  cleaned up constructors,
  1.  renamed InputRealType to RType,
  1.  made comparison operators hidden friends, and
  1.  moved volatile member overloads to the end of the class.

### CODE REVIEW

**Comment by crtrott** [comment]
Ok first of all clang-format wasn't applied. Second: @dsunder are you gonna address these things?

**Comment by dalg24** [comment]
> I agree with David's comments but I'd be fine merging as is if you rebase onto `develop` to get the format fixes from #2258
> 
> More nit-picking while we are at it:
> 
> * `KOKKOS_INLINE_FUNTION` -> `KOKKOS_FUNCTION` since all member functions are defined within the class body
> * some can be marked `constexpr`
> 
> An observation: I thought of `Kokkos::complex` as a substitute for `std::complex` until it is supported on the device. The alignment makes that one might want to prefer the class implemented in the `Kokkos::` namespace and that we are not just going to make an alias (I expect that it is possible with C++20 now that everything is constexpr). If I am going to request any change I would rather have you update the class description than attempt to make the implementation perfect.

ping @dsunder @dhollman

**Comment by dhollman** [comment]
`KOKKOS_INLINE_FUNCTION` sort of means more than just `inline` (and even if it didn't, defining the functions in the class body doesn't change how the compiler interacts with the `inline` keyword with respect to its inlining budget).  `KOKKOS_FUNCTION` is only for cases that can be handled with relocatable device code, so unless you're dealing with RDC, you should use `KOKKOS_INLINE_FUNCTION` instead of `KOKKOS_FUNCTION`.

> The alignment makes that one might want to prefer the class implemented in the Kokkos:: namespace and that we are not just going to make an alias

Sure, that's fine.  We can update the docs to reflect that.

**Comment by crtrott** [comment]
Ok this is more than we asked for, but it probably is an improvement over where we were so I merge it/.

**Comment by crtrott** [comment]
Wait @dhollman why did you add the [WIP] again? Are you planning to actively work on this today?

**Comment by dhollman** [comment]
I can remove it.  The changes I suggested still haven't been made, so that's why I marked it [WIP], but they're not absolutely necessary.  I can quickly make them today, or if I don't get around to it let's go ahead and merge this evening

**Comment by dhollman** [comment]
Also, the problem with me updating this is that I can't push to `dsunder/kokkos`, so I can't make these changes without opening a new pull request...

That, to me, seems like a vote in favor of using some shared repository as the source for pull requests rather than personal forks

**Comment by masterleinad** [comment]
> Also, the problem with me updating this is that I can't push to `dsunder/kokkos`, so I can't make these changes without opening a new pull request...

Isn't "Allow edits from maintainers" turned on by default?

**Comment by crtrott** [comment]
@dhollman  I believe you should be able to push to Dans branch if he opened a pull request from it to Kokkos/kokkos. Can you try?

File: core/src/Kokkos_Complex.hpp [hunk 1]
```diff
@@ -58,286 +58,194 @@ namespace Kokkos {
...
   }
 
-  template <typename InputRealType>
-  KOKKOS_INLINE_FUNCTION complex<RealType>& operator+=(
-      const InputRealType& src) {
-    static_assert(std::is_convertible<InputRealType, RealType>::value,
-                  "InputRealType must be convertible to RealType");
+  template <typename RType>
+  KOKKOS_INLINE_FUNCTION complex& operator+=(const RType& src) {
+    static_assert(std::is_convertible<RType, RealType>::value,
```
[the remark this answers is not in the document]
Comment (dhollman) [reply core/src/Kokkos_Complex.hpp]: we can just leave this as it is for now, I guess.

**Comment by dhollman** [comment]
Okay, this should be ready now from my perspective

**Comment by crtrott** [comment]
Can we rebase this so we don't have 68 files diff?

**Comment by dhollman** [comment]
ummm... I thought I did. The only one that differs in my repo is `Kokkos_Complex.hpp`. I can't rebase Dan's repo, I don't think.

**Comment by dhollman** [comment]
oh, never mind, I guess I can rebase that way. Should be good now

**Comment by crtrott** [comment]
I just rebased

**Comment by crtrott** [comment]
I think this is now properly rebased.

**Pull request merged**